vowpalwabbit/options.h

```diff
@@ -20,69 +20,103 @@ namespace VW
-
-  std::string m_name;
-  size_t m_type_hash;
-  std::string m_help = "";
-  std::string m_short_name = "";
-  bool m_keep = false;
-  bool m_necessary = false;
-  bool m_allow_override = false;
-
-  virtual ~base_option() = default;
-};
-
+struct base_option;
+
+// option_builder decouples the specific type of the option and the interface
+// for building it. It handles generating a shared_ptr at the end of the
+// base_option type which the options framework takes as input.
+// Therefore, T must satisfy:
+// - Inherit from base_option
+// - Have a shared_ptr field called m_default_value
+// - Statically expose a type value_type
+// - For nearly all purposes T should be typed_option<value_type> or a subclass
+//   of this.
 template <typename T>
-struct typed_option : base_option
+struct option_builder
 {
-  typed_option(const std::string& name, T& location) : base_option(name, typeid(T).hash_code()), m_location{location} {}
-
-  static size_t type_hash() { return typeid(T).hash_code(); }
+  template <typename... Args>
+  option_builder(Args&&... args) : m_option_obj(std::forward<Args>(args)...)
+  {
+  }
 
-  typed_option& default_value(T value)
+  option_builder& default_value(typename T::value_type value)
   {
-    m_default_value = std::make_shared<T>(value);
+    m_option_obj.m_default_value = std::make_shared<typename T::value_type>(value);
     return *this;
   }
 
-  bool default_value_supplied() const { return m_default_value.get() != nullptr; }
-
-  T default_value() const { return m_default_value ? *m_default_value : T(); }
-
-  typed_option& short_name(const std::string& short_name)
+  option_builder& short_name(const std::string& short_name)
   {
-    m_short_name = short_name;
+    m_option_obj.m_short_name = short_name;
     return *this;
   }
 
-  typed_option& help(const std::string& help)
+  option_builder& help(const std::string& help)
   {
-    m_help = help;
+    m_option_obj.m_help = help;
     return *this;
   }
 
-  typed_option& keep(bool keep = true)
+  option_builder& keep(bool keep = true)
   {
-    m_keep = keep;
+    m_option_obj.m_keep = keep;
     return *this;
   }
 
-  typed_option& necessary(bool necessary = true)
+  option_builder& necessary(bool necessary = true)
   {
-    m_necessary = necessary;
+    m_option_obj.m_necessary = necessary;
     return *this;
   }
 
-  typed_option& allow_override(bool allow_override = true)
+  option_builder& allow_override(bool allow_override = true)
   {
-    static_assert(is_scalar_option_type<T>::value, "allow_override can only apply to scalar option types.");
-    m_allow_override = allow_override;
+    if (!is_scalar_option_type<typename T::value_type>::value)
+    { THROW("allow_override can only apply to scalar option types.") }
+    m_option_obj.m_allow_override = allow_override;
     return *this;
   }
 
+  static std::shared_ptr<base_option> finalize(option_builder&& option)

```

Comment:
I don't understand what happens to `option` after it is moved into `finalize()` since we only care about `option.m_option_obj`

Reply:
It is moved into a `shared_ptr` and returned. This preserves the concrete type while giving back the base object to be passed to be saved.

Reply:
but we could just pass option as a const ref and then move the internal object?

Reply:
We could but the `option_builder` object is invalid after that point and calling any of the builder functions would result in undefined behavior

Reply:
fair